File: python/src/livekit/plugins/ai_coustics/analyzer.py
from __future__ import annotations

import asyncio
import math
import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import Literal

import aic_sdk
import numpy as np
from opentelemetry import metrics

from livekit import rtc

from .log import log_fields, logger
from .processor import _license_key, _pcm16_to_float32

_DEFAULT_ANALYSIS_INTERVAL = 5.0
_METER = metrics.get_meter("ai-coustics-livekit-plugin")
_ANALYSIS_COUNT = _METER.create_counter(
    "ai_coustics.analyzer.analysis",
    description="Number of ai-coustics buffered audio analyses",
)
_INFERENCE_DURATION = _METER.create_histogram(
    "ai_coustics.analyzer.inference_duration",
    unit="s",
    description="Duration of ai-coustics buffered audio analysis",
)
_SCORE = _METER.create_histogram(
    "ai_coustics.analyzer.score",
    description="Audio analysis score produced by ai-coustics",
)
_METRIC_BASE_ATTRIBUTES = {"model_provider": "ai-coustics"}
_RESULT_FIELDS = (
    "risk_score",
    "speaker_reverb",
    "speaker_loudness",
    "interfering_speech",
    "noise",
    "codec_degradation",
    "packet_loss",
)


@dataclass(frozen=True)
class AnalysisEvent:
    """Metadata and SDK result emitted after one successful buffered analysis."""

    result: aic_sdk.AnalysisResult
    timestamp: float
    """Unix timestamp in seconds recorded after inference completed."""

    inference_duration: float
    """Elapsed inference time in seconds."""

    sequence: int
    model_id: str
    room_name: str | None
    participant_identity: str | None
    publication_sid: str | None


class Collector(rtc.FrameProcessor[rtc.AudioFrame]):
    """Transparent LiveKit frame processor that collects audio for an :class:`Analyzer`."""

    def __init__(
        self,
        native_collector: aic_sdk.Collector,
        *,
        reset_analyzer: Callable[[], None],
        close_analyzer: Callable[[], None],
    ) -> None:
        self._collector: aic_sdk.Collector | None = native_collector
        self._reset_analyzer = reset_analyzer
        self._close_analyzer = close_analyzer
        self._format: tuple[int, int, int] | None = None
        self._stream_info: dict[str, str] = {}
        self._has_buffered_audio = False
        self._enabled = True
        self._closed = False

    @property
    def enabled(self) -> bool:
        return self._enabled

    @enabled.setter
    def enabled(self, value: bool) -> None:
        if self._closed or value == self._enabled:
            return
        if value:
            self._reset()
        self._enabled = value

    @property
    def initialized(self) -> bool:
        return self._has_buffered_audio and self._collector is not None

    @property
    def stream_info(self) -> dict[str, str]:
        return self._stream_info.copy()

    def _on_stream_info_updated(
        self,
        *,
        room_name: str,
        participant_identity: str,
        publication_sid: str,
    ) -> None:
        self._stream_info = {
            "room_name": room_name,
            "participant_identity": participant_identity,
            "publication_sid": publication_sid,
        }
        self._reset()

    def _on_stream_info_cleared(self) -> None:
        self._stream_info = {}
        self._reset()

    def _reset(self) -> None:
        if self._closed:
            return
        self._has_buffered_audio = False
        try:
            self._reset_analyzer()
        except Exception as error:
            logger.error(
                "Analyzer: reset failed",
                extra=log_fields(
                    "analyzer",
                    error_type=type(error).__name__,
                    error_message=str(error),
                ),
                exc_info=(type(error), error, error.__traceback__),
            )

    def _process(self, frame: rtc.AudioFrame) -> rtc.AudioFrame:
        collector = self._collector
        if not self._enabled or collector is None:
            return frame

        try:
            stream_format = (
                frame.sample_rate,
                frame.num_channels,
                frame.samples_per_channel,
            )
            if self._format != stream_format:
                collector.initialize(
                    aic_sdk.ProcessorConfig(
                        sample_rate=frame.sample_rate,
                        block_size=frame.samples_per_channel,
                        variable_block_size=False,
                    )
                )
                self._format = stream_format

            samples = _pcm16_to_float32(frame.data)
            expected_samples = frame.samples_per_channel * frame.num_channels
            if samples.size != expected_samples:
                raise ValueError(
                    f"AudioFrame contains {samples.size} samples, expected {expected_samples}"
                )

            channels = samples.reshape(frame.samples_per_channel, frame.num_channels)
            mono = (
                channels[:, 0].copy()
                if frame.num_channels == 1
                else channels.mean(axis=1, dtype=np.float32)
            )
            collector.buffer(mono)
            self._has_buffered_audio = True
        except Exception:
            logger.exception(
                "Collector: failed; passing audio through",
                extra=log_fields("collector", **self._stream_info),
            )

        return frame

    def _detach(self) -> None:
        self._closed = True
        self._enabled = False
        self._format = None
        self._stream_info = {}
        self._has_buffered_audio = False
        self._collector = None

    def _close(self) -> None:
        if self._closed:
            return
        self._detach()
        self._close_analyzer()


class Analyzer(rtc.EventEmitter[Literal["analysis_result"]]):
    """Periodically analyzes audio buffered by a transparent LiveKit Collector."""

    def __init__(
        self,
        *,
        model: aic_sdk.Model,
        license_key: str | None = None,
        analysis_interval: float = _DEFAULT_ANALYSIS_INTERVAL,
        enable_metrics: bool = True,
    ) -> None:
        super().__init__()
        if not math.isfinite(analysis_interval) or analysis_interval <= 0.0:
            raise ValueError("analysis_interval must be a finite value greater than zero")

        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            raise RuntimeError(
                "ai-coustics Analyzer must be created while an event loop is running"
            ) from None

        resolved_license_key = _license_key(license_key)
        aic_sdk.set_sdk_id(8)  # type: ignore[attr-defined]
        try:
            model_id = model.get_id()
            native_collector, native_analyzer = aic_sdk.analyzer_pair(model, resolved_license_key)
        except Exception as error:
            raise RuntimeError(f"Failed to create ai-coustics Analyzer: {error}") from error

        self._native_analyzer: aic_sdk.Analyzer | None = native_analyzer
        self._model_id = model_id
        self._analysis_interval = analysis_interval
        self._enable_metrics = enable_metrics
        self._sequence = 0
        self._closed = False
        self._close_event = asyncio.Event()
        self._close_task: asyncio.Task[None] | None = None
        self.collector = Collector(
            native_collector,
            reset_analyzer=native_analyzer.reset,
            close_analyzer=self._request_close,
        )
        self._task = loop.create_task(self._analysis_loop(), name="ai-coustics-analyzer")

    async def _terminate_session(self) -> None:
        native_analyzer = self._native_analyzer
        self._native_analyzer = None
        if native_analyzer is None:
            return
        try:
            await asyncio.to_thread(native_analyzer.terminate_session)
        except Exception:
            logger.exception(
                "Analyzer: session termination failed",
                extra=log_fields("analyzer", model_name=self._model_id),
            )

    async def _analyze_once(self, native_analyzer: aic_sdk.Analyzer) -> None:
        started = time.perf_counter()
        try:
            result = await asyncio.to_thread(native_analyzer.analyze_buffered)
        except Exception:
            inference_duration = time.perf_counter() - started
            self._record_analysis_metrics(inference_duration, status="error")
            raise

        inference_duration = time.perf_counter() - started
        self._sequence += 1
        stream_info = self.collector.stream_info
        event = AnalysisEvent(
            result=result,
            timestamp=time.time(),
            inference_duration=inference_duration,
            sequence=self._sequence,
            model_id=self._model_id,
            room_name=stream_info.get("room_name"),
            participant_identity=stream_info.get("participant_identity"),
            publication_sid=stream_info.get("publication_sid"),
        )

        self._record_analysis_metrics(inference_duration, status="ok", result=result)

        try:
            self.emit("analysis_result", event)
        except Exception:
            logger.exception(
                "Analyzer: result event emission failed",
                extra=log_fields(
                    "analyzer",
                    model_name=self._model_id,
                    sequence=self._sequence,
                    **stream_info,
                ),
            )

    def _record_analysis_metrics(
        self,
        inference_duration: float,
        *,
        status: str,
        result: aic_sdk.AnalysisResult | None = None,
    ) -> None:
        if not self._enable_metrics:
            return

        try:
            attributes = {**_METRIC_BASE_ATTRIBUTES, "status": status}
            _ANALYSIS_COUNT.add(1, attributes=attributes)
            _INFERENCE_DURATION.record(inference_duration, attributes=attributes)
            if result is not None:
                for score_name in _RESULT_FIELDS:
                    _SCORE.record(
                        getattr(result, score_name),
                        attributes={**_METRIC_BASE_ATTRIBUTES, "score.name": score_name},
                    )
        except Exception:
            logger.exception(
                "Analyzer: metrics recording failed",
                extra=log_fields(
                    "analyzer",
                    model_name=self._model_id,
                    status=status,
                ),
            )

    async def _analysis_loop(self) -> None:
        native_analyzer = self._native_analyzer
        assert native_analyzer is not None
        try:
            while not self._close_event.is_set():
                try:
                    await asyncio.wait_for(
                        self._close_event.wait(), timeout=self._analysis_interval
                    )
                    break
                except asyncio.TimeoutError:
                    pass
                if not self.collector.initialized:
                    continue
                try:
                    await self._analyze_once(native_analyzer)
                except Exception:
                    logger.exception(
                        "Analyzer: buffered audio analysis failed",
                        extra=log_fields(
                            "analyzer",
                            model_name=self._model_id,
                            **self.collector.stream_info,
                        ),
                    )
        finally:
            await self._terminate_session()

    async def _finish_close(self) -> None:
        await self._task
        # Keep an idempotent fallback in case the loop exits before entering its try/finally.
        await self._terminate_session()

    def _request_close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self.collector._detach()
        self._close_event.set()
        self._close_task = asyncio.create_task(
            self._finish_close(), name="ai-coustics-analyzer-close"
        )

    async def aclose(self) -> None:
        """Stop scheduled analysis and terminate the SDK session."""

        self._request_close()
        assert self._close_task is not None
        await self._close_task

File: node/src/analyzer.ts
import type { TypedEventEmitter as TypedEmitter } from "@livekit/typed-emitter";
import { metrics } from "@opentelemetry/api";
import { EventEmitter } from "node:events";
import {
  AudioFrame,
  FrameProcessor,
  type FrameProcessorStreamInfo,
} from "@livekit/rtc-node";

import { writeLog } from "./log.js";
import { pcm16ToFloat32 } from "./processor.js";
import {
  Analyzer as AicAnalyzer,
  type AnalysisResult,
  type Model,
  setSdkId,
} from "./sdk.js";

const DEFAULT_ANALYSIS_INTERVAL_SECONDS = 5;
const OVERLAP_WARNING_INTERVAL_MS = 10_000;
const meter = metrics.getMeter("ai-coustics-livekit-plugin");
const analysisCount = meter.createCounter("ai_coustics.analyzer.analysis", {
  description: "Number of ai-coustics buffered audio analyses",
});
const inferenceDuration = meter.createHistogram(
  "ai_coustics.analyzer.inference_duration",
  {
    unit: "s",
    description: "Duration of ai-coustics buffered audio analysis",
  },
);
const score = meter.createHistogram("ai_coustics.analyzer.score", {
  description: "Audio analysis score produced by ai-coustics",
});
const metricBaseAttributes = { model_provider: "ai-coustics" } as const;
const resultFields = [
  ["risk_score", "riskScore"],
  ["speaker_reverb", "speakerReverb"],
  ["speaker_loudness", "speakerLoudness"],
  ["interfering_speech", "interferingSpeech"],
  ["noise", "noise"],
  ["codec_degradation", "codecDegradation"],
  ["packet_loss", "packetLoss"],
] as const satisfies ReadonlyArray<readonly [string, keyof AnalysisResult]>;

export interface AnalyzerOptions {
  /** Loaded ai-coustics SDK analysis model. */
  model: Model;
  licenseKey?: string;
  /** Seconds between analyses. Defaults to 5. */
  analysisInterval?: number;
  /** Record aggregate OpenTelemetry metrics. Defaults to true. */
  enableMetrics?: boolean;
}

export interface AnalysisEvent {
  readonly result: Readonly<AnalysisResult>;
  /** Unix timestamp in milliseconds recorded after inference completed. */
  readonly timestamp: number;
  /** Elapsed inference time in milliseconds. */
  readonly inferenceDuration: number;
  readonly sequence: number;
  readonly modelId: string;
  readonly roomName?: string;
  readonly participantIdentity?: string;
  readonly publicationSid?: string;
}

export type AnalyzerCallbacks = {
  analysisResult: (event: AnalysisEvent) => void;
};

function resolveLicenseKey(value?: string): string {
  const key = value || process.env.AIC_SDK_LICENSE;
  if (!key) {
    throw new Error(
      "An ai-coustics SDK license is required. Pass licenseKey or set AIC_SDK_LICENSE.",
    );
  }
  return key;
}

function errorDetail(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

/**
 * Transparent LiveKit frame processor that collects audio for an Analyzer.
 *
 * Holds the same native analyzer as its owning {@link Analyzer} but only ever calls the
 * buffering half of that API, which does not take the analyzer lock.
 */
export class Collector extends FrameProcessor<AudioFrame> {
  private nativeAnalyzer: AicAnalyzer | null;
  private readonly resetAnalyzer: () => void;
  private readonly closeAnalyzer: () => void;
  private streamFormat: [number, number, number] | null = null;
  private streamInfo: FrameProcessorStreamInfo | null = null;
  private hasBufferedAudio = false;
  private collectingEnabled = true;
  private closed = false;

  constructor(
    nativeAnalyzer: AicAnalyzer,
    resetAnalyzer: () => void,
    closeAnalyzer: () => void,
  ) {
    super();
    this.nativeAnalyzer = nativeAnalyzer;
    this.resetAnalyzer = resetAnalyzer;
    this.closeAnalyzer = closeAnalyzer;
  }

  isEnabled(): boolean {
    return this.collectingEnabled;
  }

  setEnabled(enabled: boolean): void {
    if (this.closed || enabled === this.collectingEnabled) return;
    if (enabled) this.reset();
    this.collectingEnabled = enabled;
  }

  /** True while the collector has collected some audio the analyzer can act on. */
  get initialized(): boolean {
    return (
      this.collectingEnabled && this.hasBufferedAudio && this.nativeAnalyzer !== null
    );
  }

  get currentStreamInfo(): FrameProcessorStreamInfo | null {
    return this.streamInfo ? { ...this.streamInfo } : null;
  }

  override onStreamInfoUpdated(info: FrameProcessorStreamInfo): void {
    this.streamInfo = { ...info };
    this.reset();
  }

  override onStreamInfoCleared(): void {
    this.streamInfo = null;
    this.reset();
  }

  process(frame: AudioFrame): AudioFrame {
    const native = this.nativeAnalyzer;
    if (!this.collectingEnabled || !native) return frame;

    try {
      const streamFormat: [number, number, number] = [
        frame.sampleRate,
        frame.channels,
        frame.samplesPerChannel,
      ];
      if (
        !this.streamFormat ||
        this.streamFormat[0] !== streamFormat[0] ||
        this.streamFormat[1] !== streamFormat[1] ||
        this.streamFormat[2] !== streamFormat[2]
      ) {
        native.initialize(frame.sampleRate, frame.samplesPerChannel, false);
        this.streamFormat = streamFormat;
      }

      const expectedSamples = frame.samplesPerChannel * frame.channels;
      if (frame.data.length !== expectedSamples) {
        throw new Error(
          `AudioFrame contains ${frame.data.length} samples, expected ${expectedSamples}`,
        );
      }

      const samples = pcm16ToFloat32(frame.data);
      const mono = new Float32Array(frame.samplesPerChannel);
      if (frame.channels === 1) {
        mono.set(samples);
      } else {
        for (let sample = 0; sample < frame.samplesPerChannel; sample += 1) {
          let sum = 0;
          for (let channel = 0; channel < frame.channels; channel += 1) {
            sum += samples[sample * frame.channels + channel]!;
          }
          mono[sample] = sum / frame.channels;
        }
      }
      native.buffer(mono);
      this.hasBufferedAudio = true;
    } catch (error) {
      writeLog(
        "error",
        "collector",
        "failed; passing audio through",
        this.streamInfo ?? {},
        error,
      );
    }

    return frame;
  }

  private reset(): void {
    if (this.closed) return;
    this.hasBufferedAudio = false;
    try {
      this.resetAnalyzer();
    } catch (error) {
      writeLog(
        "error",
        "analyzer",
        "reset failed",
        { errorMessage: errorDetail(error) },
        error,
      );
    }
  }

  detach(): void {
    this.closed = true;
    this.collectingEnabled = false;
    this.streamFormat = null;
    this.streamInfo = null;
    this.hasBufferedAudio = false;
    this.nativeAnalyzer = null;
  }

  close(): void {
    if (this.closed) return;
    this.detach();
    this.closeAnalyzer();
  }
}

/** Owns an SDK analyzer pair and periodically reports analysis of collected room audio. */
export class Analyzer extends (EventEmitter as new () => TypedEmitter<AnalyzerCallbacks>) {
  readonly collector: Collector;

  private nativeAnalyzer: AicAnalyzer | null;
  private readonly timer: ReturnType<typeof setInterval>;
  private readonly modelId: string;
  private readonly enableMetrics: boolean;
  private readonly analysisIntervalMs: number;
  private analysisInFlight: Promise<void> | null = null;
  private teardown: Promise<void> | null = null;
  private skippedAnalyses = 0;
  private lastOverlapWarning: number | null = null;
  private sequence = 0;
  private closed = false;

  constructor(options: AnalyzerOptions) {
    super();
    const analysisInterval =
      options.analysisInterval ?? DEFAULT_ANALYSIS_INTERVAL_SECONDS;
    if (!Number.isFinite(analysisInterval) || analysisInterval <= 0) {
      throw new Error("analysisInterval must be a finite value greater than zero");
    }

    setSdkId(9);
    let nativeAnalyzer: AicAnalyzer;
    try {
      this.modelId = options.model.getId();
      nativeAnalyzer = new AicAnalyzer(
        options.model,
        resolveLicenseKey(options.licenseKey),
      );
    } catch (error) {
      throw new Error(`Failed to create ai-coustics Analyzer: ${errorDetail(error)}`, {
        cause: error,
      });
    }

    this.nativeAnalyzer = nativeAnalyzer;
    this.enableMetrics = options.enableMetrics ?? true;
    this.collector = new Collector(
      nativeAnalyzer,
      // Stream changes are rare and off the audio path, so reset stays synchronous even
      // though it waits for the analyzer lock when an analysis is in flight.
      () => nativeAnalyzer.reset(),
      () => void this.close(),
    );
    this.analysisIntervalMs = analysisInterval * 1000;
    this.timer = setInterval(() => this.scheduleAnalysis(), this.analysisIntervalMs);
    this.timer.unref?.();
  }

  private scheduleAnalysis(): void {
    const analyzer = this.nativeAnalyzer;
    if (this.closed || !analyzer || !this.collector.initialized) return;

    if (this.analysisInFlight) {
      this.skippedAnalyses += 1;
      const now = performance.now();
      if (
        this.lastOverlapWarning === null ||
        now - this.lastOverlapWarning >= OVERLAP_WARNING_INTERVAL_MS
      ) {
        this.lastOverlapWarning = now;
        writeLog("warn", "analyzer", "analysis falling behind its interval", {
          modelName: this.modelId,
          analysisIntervalMs: this.analysisIntervalMs,
          skippedAnalyses: this.skippedAnalyses,
          ...(this.collector.currentStreamInfo ?? {}),
        });
      }
      return;
    }

    const tracked = this.analyze(analyzer).finally(() => {
      if (this.analysisInFlight === tracked) this.analysisInFlight = null;
    });
    this.analysisInFlight = tracked;
  }

  /** Runs one analysis on a libuv worker thread. Never rejects; failures are logged. */
  private async analyze(analyzer: AicAnalyzer): Promise<void> {
    const started = performance.now();
    try {
      const nativeResult = await analyzer.analyzeAsync();
      const elapsed = performance.now() - started;
      const result = Object.freeze({ ...nativeResult });
      this.sequence += 1;
      const streamInfo = this.collector.currentStreamInfo;
      const event = Object.freeze({
        result,
        timestamp: Date.now(),
        inferenceDuration: elapsed,
        sequence: this.sequence,
        modelId: this.modelId,
        ...(streamInfo ?? {}),
      }) satisfies AnalysisEvent;

      this.recordMetrics(elapsed, "ok", result);
      try {
        this.emit("analysisResult", event);
      } catch (error) {
        writeLog(
          "error",
          "analyzer",
          "result event emission failed",
          { modelName: this.modelId, sequence: this.sequence, ...(streamInfo ?? {}) },
          error,
        );
      }
    } catch (error) {
      this.recordMetrics(performance.now() - started, "error");
      writeLog(
        "error",
        "analyzer",
        "buffered audio analysis failed",
        { modelName: this.modelId, ...(this.collector.currentStreamInfo ?? {}) },
        error,
      );
    }
  }

  private recordMetrics(
    inferenceDurationMs: number,
    status: "ok" | "error",
    result?: Readonly<AnalysisResult>,
  ): void {
    if (!this.enableMetrics) return;

    try {
      const attributes = { ...metricBaseAttributes, status };
      analysisCount.add(1, attributes);
      inferenceDuration.record(inferenceDurationMs / 1000, attributes);
      if (result) {
        for (const [scoreName, property] of resultFields) {
          score.record(result[property], {
            ...metricBaseAttributes,
            "score.name": scoreName,
          });
        }
      }
    } catch (error) {
      writeLog(
        "error",
        "analyzer",
        "metrics recording failed",
        { modelName: this.modelId, status },
        error,
      );
    }
  }

  /**
   * Stops scheduled analysis and releases the SDK session.
   *
   * Resolves once any in-flight analysis has settled. Teardown waits for it because
   * `terminateSession()` and `dispose()` block on the analyzer lock. Safe to call without
   * awaiting, and repeated calls return the same promise.
   */
  close(): Promise<void> {
    if (this.teardown) return this.teardown;
    this.closed = true;
    clearInterval(this.timer);
    this.collector.detach();
    const analyzer = this.nativeAnalyzer;
    this.nativeAnalyzer = null;
    this.teardown = (this.analysisInFlight ?? Promise.resolve()).then(() => {
      if (analyzer) this.release(analyzer);
    });
    return this.teardown;
  }

  private release(analyzer: AicAnalyzer): void {
    try {
      analyzer.terminateSession();
    } catch (error) {
      writeLog(
        "error",
        "analyzer",
        "session termination failed",
        { modelName: this.modelId, errorMessage: errorDetail(error) },
        error,
      );
    }
    try {
      analyzer.dispose();
    } catch (error) {
      writeLog(
        "error",
        "analyzer",
        "native disposal failed",
        { modelName: this.modelId, errorMessage: errorDetail(error) },
        error,
      );
    }
  }
}
